### ISSUE TITLE
Inconsistent behavior - __arglist in local methods

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#12889]
Test code included at the bottom of this report.

Local methods can be declared with an __arglist parameter and can be invoked with an __arglist argument but an __arglist parameter cannot be used inside a local method.

Also, __arglist parameters'  behavior is not consistent across lambda expressions, anonymous methods, local methods and regular methods.

In lambda expressions __arglist keyword is not parsed at all.
In anonymous methods __arglist is parsed but leads to a compilation error.
In regular methods __arglist keyword is always parsed but leads to compilation errors if used in an iterator or async method.
In local methods __arglist keyword is parsed, but cannot be used. It also leads to a compilation error if used in an iterator local method (it is consistent with regular methods) but does not lead to an error if used in async local method (it is not consistent with regular methods)

Compilation errors caused by usages of __arglist parameters are misleading. Any usage of an __arglist parameter in a local method causes CS4013 error which states that "Instance of type 'RuntimeArgumentHandle' cannot be used inside an anonymous function, query expression, iterator block or async method". However local methods are not anonymous and the error can occur in a non-iterator, non-async local method. If the error occurs in an async local method it may take considerable amount of time to find out that real problem isn't related at all to the async keyword (contrary to the error description)

Due to inability to use it declaring an __arglist parameter in a local method and passing an __arglist argument to a local method invocation is meaningless. __arglist parameter declaration itself should lead to a compilation error in local method.
However, I personally prefer them not being parsed at all in a local method signature. In this case an __arglist keyword could provide a hint whether the method should be parsed as local or regular method (the problem was described in details here: https://github.com/dotnet/roslyn/issues/8645#issuecomment-184503344 )

![__arglist](https://cloud.githubusercontent.com/assets/5566206/17360864/3f8b27ec-5976-11e6-86b8-9b5172c8e2de.png)

``` cs
    using System;
    using System.Collections.Generic;
    using System.Threading.Tasks;

    static class C
    {
        public static void Main() { }

        public static void Method(__arglist)
        {
            Func<int> lambda = (__arglist) => 5;

            Func<int> method = delegate(__arglist)
            {
                return 5;
            };

            void LocalMethod(__arglist) { }

            LocalMethod(__arglist(1, 2));

            void LocalMethod2(__arglist)
            {
                var args2 = new ArgIterator(__arglist);
            }

            async Task LocalMethod3(__arglist)
            {

            }

            IEnumerable<int> LocalMethod4(__arglist)
            {
                yield return 0;
            }
        }

        public static async Task Method2(__arglist)
        {
        }
    }
```

**Comment by TessenR** [comment]
Same problem here:
```cs
public class C<T> {
    public void M<T>() {
        void M<T>(__arglist){} // no errors
    }
    public void M2(__arglist) {} // error CS0224: A method with vararg cannot be generic, be in a generic type, or have a params parameter
}
```

**Comment by jaredpar** [comment]
Looks like this is a dupe of #10765
